Add Permissions API sample

addyosmani opened the pull request:
r: @PaulKinlan @jeffposnick

addyosmani committed:
Add Permissions API sample

addyosmani committed:
Add midi/midi-sysex permissions to example

gauntface commented on permissions/index.html:
```diff
@@ -0,0 +1,165 @@
+
+    <!-- // [START code-block] -->
+
+    <h3>Permissions</h3>
```
Looking at the enum on the spec, the permission names are "geolocation", "notiications", "push" and "midi".

Is it just the case that push isn't implemented in Chrome yet?

https://w3c.github.io/permissions/#h-permission-registry

addyosmani replied:
Latest details on what was implemented https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/exHjHQzY6m0, however I haven't been able to get push notifications working reliably with the API in Canary or ToT.

addyosmani replied:
Got it working locally against push. Will try to add commits to expand the sample to cover.

addyosmani committed:
Permissions: add Push Notifications sample

jeffposnick commented on permissions/index.html:
```diff
@@ -0,0 +1,198 @@
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+
+    <meta name="description" content="Sample illustrating the use of the Permissions aPI.">
```
aPI :arrow_right: API

addyosmani committed:
Permissions: typo correction

jeffposnick commented on permissions/index.html:
```diff
@@ -0,0 +1,198 @@
+        getNotBtn.addEventListener('click', function() {
+          Notification.requestPermission(function(result) {
+            if (result === 'denied') {
+              log('Permission wasn\'t granted. Allow a retry.');
```
What does "Allow a retry." mean?

addyosmani replied:
Dropping. There was some quirky behaviour during testing that suggested retries were possible, but I've confirmed in latest this isn't the case.

jeffposnick commented on permissions/index.html:
```diff
@@ -0,0 +1,198 @@
+      }
+
+      init();
+    })(document);
```
You're passing `document` in to the IIFE but not assigning it to a parameter, so I don't think there's any reason to pass it in.

addyosmani replied:
Good point. I should have dropped this earlier. Ditching in the next set of commits.

jeffposnick commented on permissions/index.html:
```diff
@@ -0,0 +1,198 @@
+          var update = function() { updatePermission('midi', p.status); };
+          update();
+
+          p.onchange = update;
```
In a few of these you use the `p.onchange = update;` formulation, and in a few others you use

```
p.onchange = function() {
  updatePermission('notifications', p.status);
};
```

Is there a reason that some go with one vs. the other?

In general, what about using

```
p.onchange = function() {
  updatePermission('notifications', this.status);
};
```

for all of them, switching to using `this` instead of the closure over `p`?

gauntface replied:
+1 for this.status.

jeffposnick commented:
Thanks for working on this!

Correct me if I'm wrong, but I don't believe you need any of the service worker bits. My understanding is that there's only one Notification-related permission (not a separate Push Notification one) and it doesn't look like you need to have a service worker registered in order to query its status or request access to it: http://jsbin.com/tajegozuyu/2/edit

There's no way that the service worker is going to trigger any notifications anyway (you need a real registration with a push server like GCM for that to happen), so adding in those bits just complicates the sample. The `manifest.json` can be left out, too.

Also, if @gauntface is going to write an accompanying updates.html5rocks.com article, then it may help to put in some of those code-snippet comments around the portions that could be pulled out and included inline.

gauntface commented:
I was the one who nudged the push permission - at the moment notification permission grants you push notification. This may not be the case in the future and I think this was largely to unblock push api from waiting on the permissions API.

Notifications and push are both referenced explicitly in the API doc.

gauntface commented:
Second look - yeah not sure if the SW bits are needed however - if it's a requirement for the push demo . . . . not sure what is best.

addyosmani committed:
Permissions: address feedback

addyosmani commented:
Atm, Matt is correct in that both `push` and `notifications` can be queried based on the spec. 

For the sake of simplicity, I've dropped push notifications from the sample (as well as the SW code). I've also updated to address the comments on-thread.

PTAL.

addyosmani committed:
Permissions: p.status -> this.status

gauntface commented:
LGTM

jeffposnick commented:
LGTM++.

I'm assuming the separate `push` permission is more forward thinking, since I've never run into a `push` permission prompt in Chrome 43. I can see it making sense as a standalone permission if/when push is decoupled from notifications, e.g. in the background sync use case.

addyosmani commented:
\o/

gauntface commented:
I think that is fair (ish) to say on regarding the permission.
